docs/source/contribution/cpu_features.rst

```diff
@@ -0,0 +1,214 @@
+
+.. highlight:: cpp
+
+CPU Features Dispatching
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+For each algorithm oneDAL provides several code paths for x86-64-compatibe instruction
+set architectures.
+
+Following architectures are currently supported:
+
+- Intel |reg| Streaming SIMD Extensions 2 (Intel |reg| SSE2)
+- Intel |reg| Streaming SIMD Extensions 4.2 (Intel |reg| SSE4.2)
+- Intel |reg| Advanced Vector Extensions 2 (Intel |reg| AVX2)
+- Intel |reg| Advanced Vector Extensions 512 (Intel |reg| AVX-512)
+
+The particular code path is chosen at runtime based on the underlying hardware characteristics.
+
+This chapter describes how the code is organized to support this variety of instruction sets.
+
+Algorithm Implementation Options
+********************************
+
+In addition to the instruction set architectures, an algorithm in oneDAL may have various
+implementation options. Below is a description of these options to help you better understand
+the oneDAL code structure and conventions.
+
+Computational Tasks
+-------------------
+
+An algorithm might have various tasks to compute. The most common options are:
+
+- `Classification <https://oneapi-src.github.io/oneDAL/onedal/glossary.html#term-Classification>`_,
+- `Regression <https://oneapi-src.github.io/oneDAL/onedal/glossary.html#term-Regression>`_.
+
+Computational Stages
+--------------------
+
+An algorithm might have ``training`` and ``inference`` computaion stages aimed
+to train a model on the input dataset and compute the inference results respectively.
+
+Computational Methods
+---------------------
+
+An algorithm can support several methods for the same type of computations.
+For example, kNN algorithm supports
+`brute_force <https://oneapi-src.github.io/oneDAL/onedal/algorithms/nearest-neighbors/knn.html#knn-t-math-brute-force>`_
+and `kd_tree <https://oneapi-src.github.io/oneDAL/onedal/algorithms/nearest-neighbors/knn.html#knn-t-math-kd-tree>`_
+methods for algorithm training and inference.
+
+Computational Modes
+-------------------
+
+oneDAL can provide several computaional modes for an algorithm.
+See `Computaional Modes <https://oneapi-src.github.io/oneDAL/onedal/programming-model/computational-modes.html>`_
+chapter for details.
+
+Folders and Files
+*****************
+
+Consider you are working on some algorithm ``Abc`` in oneDAL.
+
+The part of the implementation of this algorithms that is running on CPU should be located in
+`cpp/daal/src/algorithms/abc` folder.
+
+Consider it provides:
+
+- ``classification`` and ``regression`` learning tasks;
+- ``training`` and ``inference`` stages;
+- ``method1`` and ``method2`` for the ``training`` stage and only ``method1`` for ``inference`` stage;
+- only batch computational mode.
+
+Then the `cpp/daal/src/algorithms/abc` folder should contain at least the following files:
+
+::
+
+  cpp/daal/src/algorithms/abc/
+    |-- abc_classification_predict_method1_batch_fpt_cpu.cpp
```

Comment:
What does "fpt" stand for BTW?

Reply:
"floating point type" I guess.

Reply:
Thanks. Perhaps it could be added as a small comment.

Reply:
The description was added in '*_fpt_cpu.cpp' subsection.